source/server/listener_impl.cc

```diff
@@ -536,5 +628,80 @@ void ListenerImpl::setSocketFactory(const Network::ListenSocketFactorySharedPtr&
   socket_factory_ = socket_factory;
 }
 
+bool ListenerImpl::supportUpdateFilterChain(const envoy::config::listener::v3::Listener& config,
+                                            bool worker_started) {
+  // If a listener is in active list but stopped, the in place update will fail. The only known case
+  // is that listener is stopped by admin interface. It is ok if that admin method is called only
+  // prior to stop envoy. Otherwise we need to track the stopping state and execute full listener
+  // update.
+
+  if (!Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.listener_in_place_filterchain_update")) {
+    return false;
+  }
+
+  // The in place update execution flow is to replace the active listener. worker_started is the
+  // sufficient condition and is very close to the necessary condition.
+  if (!worker_started) {
+    return false;
+  }
+
+  // Currently we only support TCP filter chain update.
+  if (Network::Utility::protobufAddressSocketType(config_.address()) !=
+          Network::Address::SocketType::Stream ||
+      Network::Utility::protobufAddressSocketType(config.address()) !=
+          Network::Address::SocketType::Stream) {
+    return false;
+  }
+
+  // Guard the filter_chains[0] access.
+  if (config.filter_chains_size() == 0) {
+    return false;
+  }
+
+  // See buildProxyProtocolListenerFilter().
+  if (PROTOBUF_GET_WRAPPED_OR_DEFAULT(config_.filter_chains()[0], use_proxy_proto, false) ^
+      PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.filter_chains()[0], use_proxy_proto, false)) {
+    return false;
+  }
+
+  // See buildTlsInspectorListenerFilter().
+  if (needTlsInspector(config_) ^ needTlsInspector(config)) {
+    return false;
+  }
+  return ListenerMessageUtil::filterChainOnlyChange(config_, config);
+}
+
+ListenerImplPtr
+ListenerImpl::newListenerWithFilterChain(const envoy::config::listener::v3::Listener& config,
+                                         bool workers_started, uint64_t hash) {
+  return std::make_unique<ListenerImpl>(
```

Comment:
I think this variant of the constructor can be private?

Reply:
Right. The added constructor could be and should be private